### ISSUE TITLE
secure mode fails when image url has parameters

### ISSUE THREAD

**Issue by Enalmada** [thumbor/thumbor#1064]
Thumbor 6.4.2 on centos using pollexor to generate urls.  Unsecure everything works, turning on secure causes any image urls with parameters to fail and I have found a hint that may lead to the solution...

gravatar image with parameters anyone can test with
`secure.gravatar.com/avatar/a432db53989289d4f251c52c72ff2bd3?d=identicon&r=g&s=140
`

unsafe unencoded feteches gravatar without parameters
`localhost:8888/unsafe/48x48/smart/secure.gravatar.com/avatar/a432db53989289d4f251c52c72ff2bd3?d=identicon&r=g&s=140`

I expected pollexor to encode for me so I encoded the url manually and that works again
`localhost:8888/unsafe/48x48/smart/secure.gravatar.com%2Favatar%2Fa432db53989289d4f251c52c72ff2bd3%3Fd%3Didenticon%26r%3Dg%26s%3D140`

turning on secure of this encoded url fails
`localhost:8888/0tvjieFieX3uBi3T4J0TgVjicWs=/48x48/smart/secure.gravatar.com%2Favatar%2Fa432db53989289d4f251c52c72ff2bd3%3Fd%3Didenticon%26r%3Dg%26s%3D140`
```
2018-05-08 18:09:44 thumbor:WARNING Malformed URL: /0tvjieFieX3uBi3T4J0TgVjicWs=/48x48/smart/secure.gravatar.com/avatar/a432db53989289d4f251c52c72ff2bd3%3Fd=identicon&r=g&s=140
2018-05-08 18:09:44 tornado.access:WARNING 400 GET /0tvjieFieX3uBi3T4J0TgVjicWs=/48x48/smart/secure.gravatar.com/avatar/a432db53989289d4f251c52c72ff2bd3%3Fd=identicon&r=g&s=140 (184.187.165.208) 1.40ms
```

The interesting thing is that with secure on, it looks like the ? isn't getting decoded in the error log above.  So as a test I manually encode the ? and only the ? and then it works:
```
imageString="secure.gravatar.com/avatar/a432db53989289d4f251c52c72ff2bd3?d=identicon&r=g&s=140"
Thumbor.create("http://localhost:8888", key)
thumbor.buildImage(imageString.replace("?", "%3F"))
```
`localhost:8888/mMSk9wnUlLFvGwqyNextg117Tes=/48x48/smart/secure.gravatar.com/avatar/a432db53989289d4f251c52c72ff2bd3%3Fd=identicon&r=g&s=140`

`2018-05-09 00:09:32 tornado.access:INFO 200 GET /mMSk9wnUlLFvGwqyNextg117Tes=/48x48/smart/secure.gravatar.com/avatar/a432db53989289d4f251c52c72ff2bd3%3Fd=identicon&r=g&s=140 (184.187.165.208) 2.06ms`

Note that it also works if I pass the secure hash of question mark only encoded along with a full encoded image url:
localhost:8888/mMSk9wnUlLFvGwqyNextg117Tes=/48x48/smart/secure.gravatar.com%2Favatar%2Fa432db53989289d4f251c52c72ff2bd3%3Fd%3Didenticon%26r%3Dg%26s%3D140
`2018-05-09 00:25:59 tornado.access:INFO 200 GET /mMSk9wnUlLFvGwqyNextg117Tes=/48x48/smart/secure.gravatar.com/avatar/a432db53989289d4f251c52c72ff2bd3%3Fd=identicon&r=g&s=140 (184.187.165.208) 2.02ms`

So bottom line, unsafe works with everything encoded, secure mode only works when question mark is encoded.  I am hoping that knowing that secure works if only question is encoded helps identify what thumbor might be doing wrong in secure mode. 

Is this a bug in thumbor comparing hash to prematurely decoded url parameters?

**Comment by caeugusmao** [comment]
Hi @Enalmada,

We reproduce this warning, and It's right. According to [RFC3986, section2.4](https://tools.ietf.org/html/rfc3986#section-2.4).

Before you try to generate the thumbor url, you should encode your url first.

**Comment by Enalmada** [comment]
I really appreciate you taking the time to do that.  I guess pollexor client may be giving me the wrong hash for parameterized urls...having the above hash will help me debug what is going on.

**Comment by SHUFIL** [comment]
> Yes, for me It works. I followed the steps below:
> 
> 1 - I encoded this url http://secure.gravatar.com/avatar/a432db53989289d4f251c52c72ff2bd3?d=identicon&r=g&s=14000
> 
> 2 - I got this result: http%3A%2F%2Fsecure.gravatar.com%2Favatar%2Fa432db53989289d4f251c52c72ff2bd3%3Fd%3Didenticon%26r%3Dg%26s%3D14000
> 
> 3 - After thumborize the url, It worked for me: http://localhost:8888/Jps1sBDZgTOYKsvGIXcWz1DDv3c=/48x48/smart/http%3A%2F%2Fsecure.gravatar.com%2Favatar%2Fa432db53989289d4f251c52c72ff2bd3%3Fd%3Didenticon%26r%3Dg%26s%3D14000
> 
> I used this thumbor_key for test: my-security-key-file

How you encode this url, i have  encode using this web-sie "https://meyerweb.com/eric/tools/dencoder/" , but not at all working, but your full url only working, why
